src/lib/V3DutchDecayLib.sol

```diff
@@ -0,0 +1,145 @@
+    using FixedPointMathLib for uint256;
+    using {sub} for uint256;
+
+    /// @notice locates the current position on the curve and calculates the decay
+    /// @param curve The curve to search
+    /// @param startAmount The absolute start amount
+    /// @param decayStartBlock The absolute start block of the decay
+    function decay(V3Decay memory curve, uint256 startAmount, uint256 decayStartBlock)
+        internal
+        view
+        returns (uint256 decayedAmount)
+    {
+        // mismatch of relativeAmounts and relativeBlocks
+        if(curve.relativeAmounts.length > 16) {
+            revert InvalidDecayCurve();
+        }
+
+        // handle current block before decay or no decay
+        if (decayStartBlock >= block.number || curve.relativeAmounts.length == 0) {
+            return startAmount;
+        }
+        Uint16Array relativeBlocks = fromUnderlying(curve.relativeBlocks);
+        uint16 blockDelta = uint16(block.number - decayStartBlock);
+        // Special case for when we need to use the decayStartBlock (0)
+        if (relativeBlocks.getElement(0) > blockDelta) {
+            return linearDecay(0, relativeBlocks.getElement(0), blockDelta, startAmount, startAmount.sub(curve.relativeAmounts[0]));
+        }
+        // the current pos is within or after the curve
+        uint16 prev;
+        uint16 next;
+        (prev, next) = locateCurvePosition(curve, blockDelta);
+        uint256 lastAmount = startAmount.sub(curve.relativeAmounts[prev]);
+        uint256 nextAmount = startAmount.sub(curve.relativeAmounts[next]);
+        return linearDecay(relativeBlocks.getElement(prev), relativeBlocks.getElement(next), blockDelta, lastAmount, nextAmount);
+    }
+
+    /// @notice Locates the current position on the curve using a binary search
+    /// @param curve The curve to search
+    /// @param currentRelativeBlock The current relative position
+    /// @return prev The relative block before the current position
+    /// @return next The relative block after the current position
+    function locateCurvePosition(V3Decay memory curve, uint16 currentRelativeBlock)
+        internal
+        pure
+        returns (uint16 prev, uint16 next)
+    {
+        Uint16Array relativeBlocks = fromUnderlying(curve.relativeBlocks);
+        prev = 0;
+        next = 0;
+        while(next < curve.relativeAmounts.length) {
+            if (relativeBlocks.getElement(next) >= currentRelativeBlock) {
+                return (prev, next);
+            }
+            prev = next;
+            next++;
+        }
+        return (next - 1, next - 1);
+    }
+
+    /// @notice returns the linear interpolation between the two points
+    /// @param startBlock The start of the decay
+    /// @param endBlock The end of the decay
+    /// @param currentBlock The current position in the decay
+    /// @param startAmount The amount of the start of the decay
+    /// @param endAmount The amount of the end of the decay
+    function linearDecay(
```

Comment:
could this be refactored to share logic with the linear decay in `DutchDecayLib`?